lib/SPIRV/SPIRVTypeScavenger.h

```diff
@@ -0,0 +1,129 @@
+// Neither the names of Intel Corporation, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the necessary logic to recover pointer types from LLVM
+// IR for the output SPIR-V file after LLVM IR completes its transition to
+// opaque pointers.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRVTYPESCAVENGER_H
+#define SPIRVTYPESCAVENGER_H
+
+#include "llvm/ADT/PointerUnion.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/ValueMap.h"
+
+using namespace llvm;
+
+/// This class allows for the recovery of pointer element types from LLVM
+/// opaque pointer types.
+class SPIRVTypeScavenger {
+  /// A representation of a pointer whose type will be determined by uses. This
+  /// will include every value that needs to be assigned the same type.
+  struct DeferredType {
+    std::vector<Value *> Values;
+  };
+
+  /// This is called when a deferred type is fixed to a known use.
+  void fixType(DeferredType &Deferred, Type *AssignedType);
+
+  /// This merges two deferred types into one deferred type.
+  void mergeType(DeferredType *A, DeferredType *B);
+
+  /// A representation of the possible states of a type internal to this pass:
+  /// it may be either
+  /// * Something with a fixed LLVM type (this is a Type *)
+  /// * A type whose pointer element type is yet unknown (DeferredType *)
+  /// * A multi-level pointer type (this is a Value *, whose type is what this
+  ///   pointer type will point to). The latter should only exist for
+  ///   pointer operands of memory operations that return ptr.
+  typedef PointerUnion<Type *, DeferredType *, Value *> DeducedType;
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, DeducedType Ty) {
+    if (auto *AsTy = dyn_cast<Type *>(Ty))
+      return OS << *AsTy;
+    else if (auto *AsDeferred = dyn_cast<DeferredType *>(Ty))
+      return OS << "deferred type for " << *AsDeferred->Values[0];
+    else if (auto *AsValue = dyn_cast<Value *>(Ty))
+      return OS << "points to " << *AsValue;
+    else
+      return OS;
+  }
+
+  /// Compute the pointer element type of a value, solely based on its
+  /// definition. The value must be pointer-valued.
+  DeducedType getTypeInternal(Value *V);
```

Comment:
Maybe clarify the name:

```suggestion
  DeducedType computePointerElementType(Value *V);
```